agnpy/time_evolution/_time_evolution_utils.py

```diff
@@ -0,0 +1,238 @@
+import numpy as np
+from astropy.units import Quantity
+from pygments.styles import vs
+from astropy.constants import m_e, c, h, e
+from agnpy import InterpolatedDistribution, EmptyDistribution
+from agnpy.time_evolution.types import BinsWithDensities, FnParams, TimeEvaluationResult
+from agnpy.utils.conversion import mec2
+from scipy.interpolate import PchipInterpolator
+
+bin_size_factor = 1e-4
+
+class DistributionToSinglePointCollapseError(Exception):
+    def __init__(self, gamma_point):
+        self.gamma_point = gamma_point
+        super().__init__(f"Unsupported state, cannot create InterpolatedDistribution - distribution collapsed to a single gamma point {gamma_point}")
+
+
+def update_bin_ub(gm_bins, mask = None):
+    """ Update bin upper bounds, but only for bins indicated by the mask
+    """
+    mask = mask if mask is not None else np.ones_like(gm_bins[0], bool)
+    gm_bins[1][mask] = gm_bins[0][mask] * (1 + bin_size_factor)
+
+
+def recalc_new_rates(gm_bins, functions, unit, dens, dens_groups, previous_rates=None, recalc_mask=None):
+    """
+    Calculates (or recalculates) the energy change rates or injection rates.
+    If the mask is provided, only unmasked elements will be recalculated, and previous_rates will be used for masked elements.
+    gm_bins - array of shape (N,) consisting of gamma values (lower bounds of gamma bins), or (2,N) consisting of lower and upper bounds of gamma bins
+    previous_rates - optional dict of arrays of shape (N,) or (2,N), consisting of most recently calculated rates corresponding to gm_bins values
+    recalc_mask - optional array of shape (N,) a mask of bins that need recalculation
+    """
+    if previous_rates is None:
+        previous_rates = {}
+    mask = np.ones((gm_bins.shape[-1]), dtype=bool) if recalc_mask is None else recalc_mask
+    gm_bins_masked = gm_bins[..., mask]
+    dens_masked = dens[mask]
+    dens_groups_masked = remap_subgroups_density(mask_to_mapping(mask), dens_groups)
+    new_rates = {}
+    for label, fn in functions.items():
+        new_rates[label] = np.zeros_like(gm_bins) * unit \
+            if previous_rates.get(label) is None \
+            else previous_rates[label].copy()
+        gm_bins_interlaced = gm_bins_masked.reshape(-1)
+        params = FnParams(gm_bins_interlaced, dens_masked, dens_groups_masked)
+        new_rates[label][..., mask] = fn(params).reshape(gm_bins_masked.shape).to(unit)
+    return new_rates
+
+
+def sum_change_rates(change_rates, shape, unit, subgroups=None):
```

Comment:
it is difficult to follow what the function does, you could add a docstring to it